File: rust/crates/syntecnia-runtime/src/daemon.rs
//! Daemon — corre agentes/servers como procesos background. Port de
//! `syntecnia/runtime/daemon.py`.
//!
//! ```text
//! syntecnia daemon start program.syn    arranca en background
//! syntecnia daemon stop   program.syn   detiene
//! syntecnia daemon status               lista todos
//! syntecnia daemon logs   program.syn   muestra logs
//! syntecnia daemon restart program.syn  reinicia
//! ```
//!
//! Se desacopla de la terminal (DETACHED_PROCESS en Windows, setsid en Unix),
//! escribe un PID file y redirige salida a un log. Archivos:
//! `~/.syntecnia/daemons/<name>/{pid,log,meta}`

use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;
use std::time::{SystemTime, UNIX_EPOCH};

fn home() -> PathBuf {
    let h = std::env::var("USERPROFILE")
        .or_else(|_| std::env::var("HOME"))
        .unwrap_or_default();
    PathBuf::from(h)
}

fn daemon_dir() -> PathBuf {
    let d = home().join(".syntecnia").join("daemons");
    let _ = fs::create_dir_all(&d);
    d
}

/// Nombre del daemon = stem del path del programa.
pub fn daemon_name(program_path: &str) -> String {
    Path::new(program_path)
        .file_stem()
        .map(|s| s.to_string_lossy().into_owned())
        .unwrap_or_else(|| program_path.to_string())
}

fn state_dir(name: &str) -> PathBuf {
    let d = daemon_dir().join(name);
    let _ = fs::create_dir_all(&d);
    d
}

fn read_pid(name: &str) -> Option<u32> {
    let pid_file = state_dir(name).join("pid");
    fs::read_to_string(pid_file).ok().and_then(|s| s.trim().parse().ok())
}

fn now_secs() -> f64 {
    SystemTime::now().duration_since(UNIX_EPOCH).map(|d| d.as_secs_f64()).unwrap_or(0.0)
}

/// ¿El proceso `pid` está vivo? (tasklist en Windows, kill(0) en Unix.)
#[cfg(windows)]
fn is_running(pid: u32) -> bool {
    let out = Command::new("tasklist")
        .args(["/FI", &format!("PID eq {}", pid)])
        .output();
    match out {
        Ok(o) => String::from_utf8_lossy(&o.stdout).contains(&pid.to_string()),
        Err(_) => false,
    }
}

#[cfg(not(windows))]
fn is_running(pid: u32) -> bool {
    // En Unix, kill -0 chequea existencia.
    Command::new("kill")
        .args(["-0", &pid.to_string()])
        .output()
        .map(|o| o.status.success())
        .unwrap_or(false)
}

/// Resultado de una operación de daemon (mensaje listo para imprimir).
pub struct DaemonResult {
    pub status: String,
    pub name: String,
    pub pid: Option<u32>,
    pub message: String,
}

/// Lanza el comando detachado del terminal con stdout/stderr → `log`.
fn spawn_detached(program_path: &str, extra: &[String], log_file: fs::File) -> std::io::Result<u32> {
    let exe = std::env::current_exe()?;
    let log_err = log_file.try_clone()?;
    let mut cmd = Command::new(exe);
    // `serve` bindea cualquier `serve on PORT` y bloquea hasta kill (mantiene vivo el
    // daemon). El oráculo usa `run --serve`; acá `serve` es el equivalente que sí
    // levanta el server bajo el modelo std::net del port.
    cmd.arg("serve");
    for a in extra {
        cmd.arg(a);
    }
    cmd.arg(program_path);
    cmd.stdin(std::process::Stdio::null())
        .stdout(std::process::Stdio::from(log_file))
        .stderr(std::process::Stdio::from(log_err));

    #[cfg(windows)]
    {
        use std::os::windows::process::CommandExt;
        // DETACHED_PROCESS (0x8) | CREATE_NEW_PROCESS_GROUP (0x200)
        cmd.creation_flags(0x0000_0008 | 0x0000_0200);
    }
    #[cfg(unix)]
    {
        use std::os::unix::process::CommandExt;
        // Nueva sesión: cerrar la terminal no mata el daemon.
        cmd.process_group(0);
    }

    let child = cmd.spawn()?;
    Ok(child.id())
}

/// Quita el prefijo extended-length de Windows (`\\?\`) que deja `canonicalize`.
fn clean_path(p: String) -> String {
    p.strip_prefix(r"\\?\").map(|s| s.to_string()).unwrap_or(p)
}

pub fn daemon_start(program_path: &str, extra: &[String]) -> DaemonResult {
    let abs = std::fs::canonicalize(program_path)
        .map(|p| clean_path(p.to_string_lossy().into_owned()))
        .unwrap_or_else(|_| program_path.to_string());
    if !Path::new(&abs).exists() {
        return DaemonResult {
            status: "error".into(),
            name: String::new(),
            pid: None,
            message: format!("File not found: {}", abs),
        };
    }
    let name = daemon_name(&abs);
    let sdir = state_dir(&name);

    if let Some(existing) = read_pid(&name) {
        if is_running(existing) {
            return DaemonResult {
                status: "already_running".into(),
                name: name.clone(),
                pid: Some(existing),
                message: format!("Daemon '{}' is already running (PID {})", name, existing),
            };
        }
    }

    let log_path = sdir.join("log");
    let mut log_file = match fs::OpenOptions::new().create(true).append(true).open(&log_path) {
        Ok(f) => f,
        Err(e) => {
            return DaemonResult {
                status: "error".into(),
                name,
                pid: None,
                message: format!("could not open log: {}", e),
            }
        }
    };
    use std::io::Write;
    let _ = writeln!(log_file, "\n--- Daemon started ---");
    let _ = log_file.flush();

    match spawn_detached(&abs, extra, log_file) {
        Ok(pid) => {
            let _ = fs::write(sdir.join("pid"), pid.to_string());
            let meta = serde_json::json!({
                "program": abs,
                "name": name,
                "pid": pid,
                "started_at": now_secs(),
                "args": extra,
            });
            let _ = fs::write(sdir.join("meta"), serde_json::to_string_pretty(&meta).unwrap_or_default());
            DaemonResult {
                status: "started".into(),
                name: name.clone(),
                pid: Some(pid),
                message: format!("Daemon '{}' started (PID {}). Logs: {}", name, pid, log_path.display()),
            }
        }
        Err(e) => DaemonResult {
            status: "error".into(),
            name,
            pid: None,
            message: format!("could not start daemon: {}", e),
        },
    }
}

fn resolve_name(name_or_path: &str) -> String {
    if name_or_path.contains('/') || name_or_path.contains('\\') || name_or_path.contains('.') {
        daemon_name(name_or_path)
    } else {
        name_or_path.to_string()
    }
}

pub fn daemon_stop(name_or_path: &str) -> DaemonResult {
    let name = resolve_name(name_or_path);
    let pid = match read_pid(&name) {
        Some(p) => p,
        None => {
            return DaemonResult {
                status: "not_found".into(),
                name: name.clone(),
                pid: None,
                message: format!("No daemon '{}' found", name),
            }
        }
    };
    if !is_running(pid) {
        let _ = fs::remove_file(state_dir(&name).join("pid"));
        return DaemonResult {
            status: "not_running".into(),
            name: name.clone(),
            pid: Some(pid),
            message: format!("Daemon '{}' is not running (stale PID {})", name, pid),
        };
    }

    #[cfg(windows)]
    {
        let _ = Command::new("taskkill").args(["/PID", &pid.to_string(), "/F"]).output();
    }
    #[cfg(not(windows))]
    {
        let _ = Command::new("kill").arg(pid.to_string()).output();
    }

    let _ = fs::remove_file(state_dir(&name).join("pid"));
    DaemonResult {
        status: "stopped".into(),
        name: name.clone(),
        pid: Some(pid),
        message: format!("Daemon '{}' stopped", name),
    }
}

pub struct DaemonInfo {
    pub name: String,
    pub pid: Option<u32>,
    pub running: bool,
    pub program: String,
    pub started_at: String,
}

pub fn daemon_status() -> Vec<DaemonInfo> {
    let mut out = Vec::new();
    let dir = daemon_dir();
    let mut entries: Vec<PathBuf> = match fs::read_dir(&dir) {
        Ok(rd) => rd.filter_map(|e| e.ok().map(|e| e.path())).filter(|p| p.is_dir()).collect(),
        Err(_) => return out,
    };
    entries.sort();
    for sdir in entries {
        let name = sdir.file_name().map(|s| s.to_string_lossy().into_owned()).unwrap_or_default();
        let pid = read_pid(&name);
        let running = pid.map(is_running).unwrap_or(false);
        let (mut program, mut started_at) = (String::new(), String::new());
        if let Ok(meta_txt) = fs::read_to_string(sdir.join("meta")) {
            if let Ok(meta) = serde_json::from_str::<serde_json::Value>(&meta_txt) {
                program = meta.get("program").and_then(|v| v.as_str()).unwrap_or("").to_string();
                started_at = meta.get("started_at").map(|v| v.to_string()).unwrap_or_default();
            }
        }
        out.push(DaemonInfo { name, pid, running, program, started_at });
    }
    out
}

pub fn daemon_logs(name_or_path: &str, lines: usize) -> String {
    let name = resolve_name(name_or_path);
    let log_path = state_dir(&name).join("log");
    match fs::read_to_string(&log_path) {
        Ok(content) => {
            let all: Vec<&str> = content.split('\n').collect();
            let start = all.len().saturating_sub(lines);
            all[start..].join("\n")
        }
        Err(_) => format!("No logs for daemon '{}'", name),
    }
}

pub fn daemon_restart(name_or_path: &str, extra: &[String]) -> DaemonResult {
    let name = resolve_name(name_or_path);
    // Recupera el path del programa desde meta.
    let mut program_path = name_or_path.to_string();
    let mut args = extra.to_vec();
    if let Ok(meta_txt) = fs::read_to_string(state_dir(&name).join("meta")) {
        if let Ok(meta) = serde_json::from_str::<serde_json::Value>(&meta_txt) {
            if let Some(p) = meta.get("program").and_then(|v| v.as_str()) {
                program_path = p.to_string();
            }
            if args.is_empty() {
                if let Some(a) = meta.get("args").and_then(|v| v.as_array()) {
                    args = a.iter().filter_map(|v| v.as_str().map(String::from)).collect();
                }
            }
        }
    }
    let _ = daemon_stop(&name);
    std::thread::sleep(std::time::Duration::from_millis(500));
    daemon_start(&program_path, &args)
}

pub fn format_status_table(statuses: &[DaemonInfo]) -> String {
    if statuses.is_empty() {
        return "No daemons found.".to_string();
    }
    let mut lines = vec!["Syntecnia Daemons:".to_string()];
    for s in statuses {
        let state = if s.running { "RUNNING" } else { "STOPPED" };
        let pid = s.pid.map(|p| p.to_string()).unwrap_or_else(|| "-".to_string());
        lines.push(format!(
            "  [{:7}] {:20} PID={:8} {}  {}",
            state, s.name, pid, s.started_at, s.program
        ));
    }
    lines.join("\n")
}

File: synsema/runtime/daemon.py
"""
Synsema Daemon — Run agents as system services.

    synsema daemon start program.syn     Start as background daemon
    synsema daemon stop program.syn      Stop a running daemon
    synsema daemon status                Show all running daemons
    synsema daemon logs program.syn      Tail the daemon logs
    synsema daemon restart program.syn   Restart a daemon

The daemon:
    - Detaches from the terminal (real fork on Unix, subprocess on Windows)
    - Writes a PID file for management
    - Redirects output to log files
    - Keeps cron jobs and agents alive
    - Auto-restarts on crash (optional)
    - Supports multiple daemons simultaneously

Files:
    ~/.synsema/daemons/<name>/pid      Process ID
    ~/.synsema/daemons/<name>/log      stdout + stderr
    ~/.synsema/daemons/<name>/meta     program path, start time, etc.

Works on Linux, Mac, and Windows. Uses os.fork on Unix,
subprocess.Popen on Windows.
"""

import os
import sys
import json
import time
import signal
import subprocess
from pathlib import Path
from typing import Optional, Dict, List


def _daemon_dir() -> Path:
    """Get the daemon state directory."""
    d = Path.home() / ".synsema" / "daemons"
    d.mkdir(parents=True, exist_ok=True)
    return d


def _daemon_name(program_path: str) -> str:
    """Derive a daemon name from the program path."""
    return Path(program_path).stem


def _daemon_state_dir(name: str) -> Path:
    d = _daemon_dir() / name
    d.mkdir(parents=True, exist_ok=True)
    return d


def _read_pid(name: str) -> Optional[int]:
    pid_file = _daemon_state_dir(name) / "pid"
    if pid_file.exists():
        try:
            return int(pid_file.read_text().strip())
        except (ValueError, OSError):
            pass
    return None


def _is_running(pid: int) -> bool:
    """Check if a process is running. Works on Linux, Mac, and Windows."""
    if os.name == "nt":
        # Windows: use tasklist
        try:
            output = subprocess.check_output(
                ["tasklist", "/FI", f"PID eq {pid}"],
                stderr=subprocess.DEVNULL,
                text=True,
            )
            return str(pid) in output
        except (subprocess.SubprocessError, FileNotFoundError):
            return False
    else:
        # Unix: signal 0 checks existence
        try:
            os.kill(pid, 0)
            return True
        except (OSError, ProcessLookupError):
            return False


def daemon_start(program_path: str, extra_args: List[str] = None,
                 auto_restart: bool = False) -> Dict:
    """
    Start a Synsema program as a background daemon.

    Returns dict with: name, pid, log_path, status
    """
    program_path = str(Path(program_path).resolve())
    if not Path(program_path).exists():
        return {"status": "error", "message": f"File not found: {program_path}"}

    name = _daemon_name(program_path)
    state_dir = _daemon_state_dir(name)

    # Check if already running
    existing_pid = _read_pid(name)
    if existing_pid and _is_running(existing_pid):
        return {
            "status": "already_running",
            "name": name,
            "pid": existing_pid,
            "message": f"Daemon '{name}' is already running (PID {existing_pid})",
        }

    # Build command
    python = sys.executable
    cmd = [python, "-m", "synsema", "run", program_path, "--serve"]
    if extra_args:
        cmd.extend(extra_args)

    # Log file
    log_path = state_dir / "log"

    # Start as detached subprocess
    log_file = open(log_path, "a")
    log_file.write(f"\n--- Daemon started at {time.strftime('%Y-%m-%d %H:%M:%S')} ---\n")
    log_file.flush()

    # Use DETACHED_PROCESS on Windows, setsid on Unix
    kwargs = {
        "stdout": log_file,
        "stderr": log_file,
        "stdin": subprocess.DEVNULL,
    }

    if os.name == "nt":
        # Windows
        kwargs["creationflags"] = subprocess.DETACHED_PROCESS | subprocess.CREATE_NEW_PROCESS_GROUP
    else:
        # Unix — start in new session so terminal close doesn't kill it
        kwargs["start_new_session"] = True

    proc = subprocess.Popen(cmd, **kwargs)

    # Write PID
    (state_dir / "pid").write_text(str(proc.pid))

    # Write metadata
    meta = {
        "program": program_path,
        "name": name,
        "pid": proc.pid,
        "started_at": time.time(),
        "started_at_human": time.strftime("%Y-%m-%d %H:%M:%S"),
        "auto_restart": auto_restart,
        "args": extra_args or [],
    }
    (state_dir / "meta").write_text(json.dumps(meta, indent=2))

    return {
        "status": "started",
        "name": name,
        "pid": proc.pid,
        "log": str(log_path),
        "message": f"Daemon '{name}' started (PID {proc.pid}). Logs: {log_path}",
    }


def daemon_stop(name_or_path: str) -> Dict:
    """Stop a running daemon."""
    name = _daemon_name(name_or_path) if "/" in name_or_path or "." in name_or_path else name_or_path

    pid = _read_pid(name)
    if not pid:
        return {"status": "not_found", "message": f"No daemon '{name}' found"}

    if not _is_running(pid):
        # Clean up stale PID
        (_daemon_state_dir(name) / "pid").unlink(missing_ok=True)
        return {"status": "not_running", "message": f"Daemon '{name}' is not running (stale PID {pid})"}

    # Send SIGTERM (graceful) then force kill if needed
    try:
        if os.name == "nt":
            # Windows: taskkill
            subprocess.run(["taskkill", "/PID", str(pid), "/F"],
                           capture_output=True)
        else:
            os.kill(pid, signal.SIGTERM)
            # Wait up to 5 seconds for graceful shutdown
            for _ in range(50):
                if not _is_running(pid):
                    break
                time.sleep(0.1)
            else:
                # Force kill (Unix only)
                if hasattr(signal, "SIGKILL"):
                    os.kill(pid, signal.SIGKILL)
    except (OSError, ProcessLookupError):
        pass

    (_daemon_state_dir(name) / "pid").unlink(missing_ok=True)

    # Log stop
    log_path = _daemon_state_dir(name) / "log"
    if log_path.exists():
        with open(log_path, "a") as f:
            f.write(f"\n--- Daemon stopped at {time.strftime('%Y-%m-%d %H:%M:%S')} ---\n")

    return {"status": "stopped", "name": name, "pid": pid, "message": f"Daemon '{name}' stopped"}


def daemon_status() -> List[Dict]:
    """Get status of all daemons."""
    results = []
    daemon_dir = _daemon_dir()
    if not daemon_dir.exists():
        return results

    for state_dir in sorted(daemon_dir.iterdir()):
        if not state_dir.is_dir():
            continue
        name = state_dir.name
        meta_file = state_dir / "meta"
        pid = _read_pid(name)

        info = {
            "name": name,
            "pid": pid,
            "running": pid is not None and _is_running(pid),
        }

        if meta_file.exists():
            try:
                meta = json.loads(meta_file.read_text())
                info["program"] = meta.get("program", "")
                info["started_at"] = meta.get("started_at_human", "")
            except:
                pass

        results.append(info)

    return results


def daemon_logs(name_or_path: str, lines: int = 50) -> str:
    """Get recent log lines for a daemon."""
    name = _daemon_name(name_or_path) if "/" in name_or_path or "." in name_or_path else name_or_path
    log_path = _daemon_state_dir(name) / "log"

    if not log_path.exists():
        return f"No logs for daemon '{name}'"

    content = log_path.read_text()
    all_lines = content.split("\n")
    recent = all_lines[-lines:]
    return "\n".join(recent)


def daemon_restart(name_or_path: str, extra_args: List[str] = None) -> Dict:
    """Restart a daemon."""
    name = _daemon_name(name_or_path) if "/" in name_or_path or "." in name_or_path else name_or_path

    # Get program path from meta
    meta_file = _daemon_state_dir(name) / "meta"
    program_path = name_or_path
    if meta_file.exists():
        try:
            meta = json.loads(meta_file.read_text())
            program_path = meta.get("program", name_or_path)
            if not extra_args:
                extra_args = meta.get("args", [])
        except:
            pass

    daemon_stop(name)
    time.sleep(0.5)
    return daemon_start(program_path, extra_args)


def format_status_table(statuses: List[Dict]) -> str:
    """Format daemon status as readable table."""
    if not statuses:
        return "No daemons found."
    lines = ["Synsema Daemons:"]
    for s in statuses:
        state = "RUNNING" if s.get("running") else "STOPPED"
        pid = s.get("pid") or "-"
        name = s.get("name", "?")
        program = s.get("program", "")
        started = s.get("started_at", "")
        lines.append(f"  [{state:7s}] {name:20s} PID={str(pid):8s} {started}  {program}")
    return "\n".join(lines)
